### ISSUE TITLE
BigDecimal/float difference between MRI and JRuby

### ISSUE THREAD

**Issue by Confusion** [jruby/jruby#5904]
### Expected Behavior
In MRI the following comparison returns false:
```
2.5.3 :001 > BigDecimal.new('.803e1') == 8.03
 => false 
```
This is probably because there is no exact floating point representation for the number 8.03 (as opposed to e.g. 8.02 and 8.04, for which the similar comparison returns `true`.

### Actual Behavior
```
jruby-9.2.8.0 :001 > BigDecimal.new('0.803e1') == 8.03
 => true 
```
This is not a problem for me, except that it was surprising that a spec passed in JRuby that I expected to fail. 

### Environment

* jruby 9.2.8.0 (2.5.3) 2019-08-20 a1ac7ff OpenJDK 64-Bit Server VM 11.0.4+11-post-Ubuntu-1ubuntu218.04.3 on 11.0.4+11-post-Ubuntu-1ubuntu218.04.3 +jit [linux-x86_64]
* Linux photon 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

**Comment by mohits** [comment]
I think that this issue should be closed since this is probably implementation dependent, and should not be something that a programmer should rely on.

**Comment by ahorek** [comment]
actually https://github.com/jruby/jruby/commit/73764ea86fda0f17827b50a19b28e1de04295ef4 didn't cause the regression.

> why you chose to pass in a MathContext with a precision here?

the idea behind this is that more accurate precision than the maximum float precision doesn't matter for a comparsion with another float.

the flaw is we're not comparing the real float representation which is inacurrate.

```
BigDecimal('.803e1') == 8.03

[BigDecimal('.803e1'), BigDecimal('.803e1')]
=> true # current behaviour

[BigDecimal(8.02999973297119140625), BigDecimal('.803e1')]
=> false # desired behaviour
```

so the value should be normalized before converting to a bigdecimal. The precision isn't a real problem.

also the way how CRuby do it seems to be complex and probably slow. The best way would be to use the fast path which works for most cases and do an additional validation step if there's a match for corner cases like this.

btw: I don't see a real use case why should anyone depend on precise comparsions between floats and bigdecimals.
But yes, it's a bug - ```BigDecimal('8.03')``` and ```Float(8.03)``` aren't the same numbers even if they look very simmilar.

**Comment by headius** [comment]
> the flaw is we're not comparing the real float representation which is inacurrate.

You are correct. I did not mean to imply that you broke anything. It is largely a bug in us following CRuby logic for the comparison but not for the "GetVpValue" logic that converts float to BigDecimal via a string.

I think we will need to diverge on the implementation of RubyBigDecimal.cmp in order to do more accurate comparisons. Truth be told, we could probably drop most of the ported logic from CRuby if we can get the JDK BigDecimal to do the equivalent work.

At this point I am not sure what the fix should be, or indeed whether a fix is necessary. These values are logically equal but physically unequal, and every resource I have looked at recommends against comparing floating-point values and BigDecimal values without explicitly opting into the rounding. In this case we are rounding to 15 digits which causes them to appear equal, while 16 would cause them to be unequal:

```
$ jruby -rbigdecimal -e "p BigDecimal(8.03, 15)"
0.803e1

$ jruby -rbigdecimal -e "p BigDecimal(8.03, 16)"
0.8029999999999999e1
```

Even if we were to skip this coercion to BigDecimal at a Ruby level, we would still need to coerce to a JDK BigDecimal in order to do the comparison, at which point we would need to choose some level of rounding.
